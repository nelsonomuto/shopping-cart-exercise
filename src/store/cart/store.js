const LOCALSTORAGE_CART_KEY = "rtg-cart"

const setCartItemsLocalStorage = cartItems => {
  localStorage.setItem(LOCALSTORAGE_CART_KEY, JSON.stringify(cartItems))
}

const getCartItemsLocalStorage = () => {
  if (!localStorage.getItem(LOCALSTORAGE_CART_KEY)) {
    setCartItemsLocalStorage([])
  }

  return JSON.parse(localStorage.getItem(LOCALSTORAGE_CART_KEY))
}

const store = {
  get: getCartItemsLocalStorage,
  set: setCartItemsLocalStorage,
}

export default store
