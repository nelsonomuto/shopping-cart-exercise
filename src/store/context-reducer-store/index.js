import React from "react"
import createPersistentStore from "./persistentStore"

export default function createContextStore(
  name,
  reducer,
  defaultState,
  options = {}
) {
  // Create a "global" context to store our redux state in
  // this is private, we abstract this away from consumer
  const StoreContext = React.createContext()
  const _store = createPersistentStore(name, defaultState)

  // Set up serializers/deserializers for the store (if you only want to
  // stash part of it, or if you want to encrypt it)
  const self = content => content
  const serialize = options.serializeStore || self
  const deserialize = options.deserializeStore || self
  const store = { clear: _store.clear }
  store.set = content => _store.set(serialize(content))
  store.get = () => deserialize(_store.get())

  // Use this to wrap the whole portion of the app that needs access to the Store
  const StoreProvider = ({ children }) => {
    const initialState = store.get()
    const [state, dispatch] = React.useReducer(reducer, initialState)
    return (
      <StoreContext.Provider value={{ state, dispatch }}>
        {children}
      </StoreContext.Provider>
    )
  }

  // Use this to access state and dispatch action, like useReducer
  // except it is available locally anywhere under the StoreProvider wrapper
  const useStore = () => {
    const { state, dispatch } = React.useContext(StoreContext)

    React.useEffect(() => {
      store.set(state)
    }, [state])

    return [state, dispatch]
  }

  return {
    store,
    useStore,
    StoreProvider,
  }
}
