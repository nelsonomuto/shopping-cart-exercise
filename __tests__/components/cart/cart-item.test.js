import { mount, shallow } from "enzyme"
import React from "react"
import CartItem from "../../../src/components/cart/cart-item"
import store from "../../../src/store/cart/store"

const mockItem = {
  sku: "fake-sku",
  title: "fake-title",
  price: 999.99,
  image: "fake-image-url",
  qty: 2,
}

const mockItem2 = {
  sku: "fake-sku2",
  title: "fake-title2",
  price: 999.99,
  image: "fake-image-url2",
  qty: 1,
}

describe("<CartItem />", () => {
  afterEach(() => {
    store.clear()
  })

  it("renders items from store in cart", () => {
    store.set([mockItem])
    const component = shallow(<CartItem item={mockItem} />)
    expect(component).toMatchSnapshot()
  })

  it("clicking decrement button updates item qty", () => {
    store.set([mockItem])
    const component = mount(<CartItem item={mockItem} />)
    const decrementButton = component.find(".cart-item-update-qty-button").at(0)
    // Trigger quantity update in store on item
    decrementButton.simulate("click")
    // Store should have been updated
    const nextItem = store.get()[0]
    expect(nextItem.qty).toBe(1)
  })

  it("clicking increment button updates item qty", () => {
    store.set([mockItem])
    const component = mount(<CartItem item={mockItem} />)
    const incrementButton = component.find(".cart-item-update-qty-button").at(1)
    // Trigger quantity update in store on item
    incrementButton.simulate("click")
    // Store should have been updated
    const nextItem = store.get()[0]
    expect(nextItem.qty).toBe(3)
  })

  it("clicking remove item removes item from store", () => {
    store.set([mockItem, mockItem2])
    const component = mount(<CartItem item={mockItem} />)
    const removeButton = component.find(".cart-item-remove-button")
    // Trigger item remove on store
    removeButton.simulate("click")
    // Store should have been updated
    const nextStore = store.get()
    expect(nextStore).toEqual([mockItem2])
  })
})
