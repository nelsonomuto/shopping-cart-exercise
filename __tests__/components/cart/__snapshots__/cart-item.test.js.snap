// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> matches snapshot 1`] = `
<div>
  <h4
    className="cart-item-title"
  >
    fake-title
  </h4>
  <div
    className="cart-item-controls"
  >
    <button
      className="cart-item-update-qty-button"
      onClick={[Function]}
    >
      -
    </button>
    <span
      className="cart-item-qty"
    />
    <button
      className="cart-item-update-qty-button"
      onClick={[Function]}
    >
      +
    </button>
    <button
      className="cart-item-remove-button"
      onClick={[Function]}
    >
      Remove Item
    </button>
  </div>
</div>
`;
